Consistent use of singleton types in manual.

tpapp opened the pull request:
Fixes #7135, see discussion there.

### Changes

- separate discussion of singleton types (as in `Base.issingletontype`) and `Type{T}`

- move discussion about singleton types to a later section so that parametric singleton types can be discussed

- add example and discussion about parametric singleton types

- add usage hint about `Type` outside dispatch ~~(TL;DR: don't)~~ example: type sharpening (thanks @tkf)

### Open questions:

- [x] is there a terminology for `Type{T}` now that we don't confuse it with singleton types? If not, do we want to introduce one, eg call it "type selectors"? 

- [x] should I mention that abstract types can't be singleton types, or is it obvious from the context?

- [x] should I expunge the remaining use of 'singleton' for `Type{T}` from *Conversion and Promotion*?

tpapp committed:
Consistent use of singleton types in manual.

Fixes #7135, see discussion there.

- separate discussion of singleton types (as in
  `Base.issingletontype`) and `Type{T}`

- move discussion about singleton types to a later section so that
  parametric singleton types can be discussed

- add example and discussion about parametric singleton types

- add usage hint about `Type` outside dispatch (TL;DR: don't)

tkf commented:
> add usage hint about `Type` outside dispatch (TL;DR: don't)

A usage (workaround) outside dispatch is to get a sharp field type.  For example, the type of `Fix1(Vector, undef)` is `Fix1{Type{Vector},UndefInitializer}` rather than `Fix1{DataType,UndefInitializer}`.  This is required for the field type to be inferable.  So, I think it'd be better to avoid stressing that `Type{T}` is _only_ for dispatch.  "`Type{T}` types" section in this PR seems to be fine as-is, though.

tkf commented on doc/src/manual/types.md:
```diff
@@ -1127,6 +1058,123 @@ must always be specified in full, this is not especially helpful, but in Julia,
 
+## [Singleton types](@id man-singleton-types)
+
+Immutable composite types with no fields are called *singletons*. Formally, if 
+`a isa T && b isa T` implies `a === b`, then `T` is a singleton type. [^2]
```
I think the spirit of this formalism is great, but if I have

```julia
julia> abstract type A end

julia> struct B <: A end
```

and I don't define any other subtype of `A`, would it imply that `A` is also a singleton type?

Maybe making this perfect requires `isconcretetype(T)`?  Although if you say "for any `a` and `b`" I guess it could mean "for any `a` and `b` of any implementable type."

> should I mention that abstract types can't be singleton types, or is it obvious from the context?

So yes, I think this helps to make this definition clearer.

tpapp replied:
Yes, I think we should stress that only concrete types can to be singletons.

tkf commented:
> is there a terminology for `Type{T}` now that we don't confuse it with singleton types? If not, do we want to introduce one, eg call it "type selectors"?

Yeah, calling `Type{T}` a singleton sounds a bit confusing since you can write something like `Type{<:Integer}`.  For the new name, "type selectors" sound reasonable to me.

> should I expunge the remaining use of 'singleton' for `Type{T}` from _Conversion and Promotion_?

:+1: from me :)

tpapp commented:
@tkf: thanks for the review. Yes, sharp field types are another use case --- if you have a small, self-contained example, I would be happy to add it to the PR if that's OK with you.

I included the usage hint because I have seen users on Discourse trying to work with `Type{Type{T}}` etc constructs and other misunderstandings. `Type` is mysterious to a newcomer as is, and each common use case should serve as an example. So I would rather just include them if feasible.

tkf commented:
Yeah, including example is a good idea.  I think a minimal example is `Some`:

```julia
struct SomeBad{T}
    value::T
end

struct SomeGood{T}
    value::T
end
SomeGood(::Type{T}) where {T} = SomeGood{Type{T}}(T)

sbad = SomeBad(Symbol)
sgood = SomeGood(Symbol)

@assert sbad isa SomeBad{DataType}
@assert sgood isa SomeGood{Type{Symbol}}

getvalue(x) = x.value

@code_warntype getvalue(sbad)  # => Body::DataType
@code_warntype getvalue(sgood)  # => Body::Type{Symbol}
```

tpapp committed:
minor changes

- remove incorrect reference to "singleton types" from the conversion
chapter

- emphasize that singleton types need to be concrete types,
specifically mention abstract types

- include a type sharpening example for `Type{T}` (thanks @tkf)

- introduce the term "type selector"

tpapp commented:
Thanks @tkf, I simplified the example somewhat as I did not want to deal with `@code_warntype` in a `jldoctest`. I also did the other changes above.

tpapp commented:
Friendly bump — if this needs to be improved, I am happy to implement any further suggested changes.

StefanKarpinski commented:
I've requested review by @JeffBezanson. I'll merge one way or another this week.

rfourquet commented on doc/src/manual/conversion-and-promotion.md:
```diff
@@ -168,13 +168,12 @@ Such a definition might look like this:
 
-The type of the first argument of this method is a [singleton type](@ref man-singleton-types),
+The type of the first argument of this method is a [`Type{T}`](@ref man-typet-type),
 `Type{MyType}`, the only instance of which is `MyType`. Thus, this method is only invoked
```
Why not put directly `Type{MyType}` within the `[]` brackets, where `Type{T}` is currently?

rfourquet commented on doc/src/manual/types.md:
```diff
@@ -1127,6 +1058,146 @@ must always be specified in full, this is not especially helpful, but in Julia,
+
+## [`Type{T}` type selectors](@id man-typet-type)
+
+For each type `T`, the `Type{T}` is an abstract parametric type whose only instance is the
```
"the `Type{T}` is ..." sounds weird, it reads "the type `T`", where in fact we mean "the `Type{T}` type". I would just remove "the":
```suggestion
For each type `T`, `Type{T}` is an abstract parametric type whose only instance is the
```

rfourquet commented on doc/src/manual/types.md:
````diff
@@ -1127,6 +1058,146 @@ must always be specified in full, this is not especially helpful, but in Julia,
+false
+```
+
+In other words, [`isa(A,Type{B})`](@ref) is true if and only if `A` and `B` are the same object
````
```suggestion
In other words, [`isa(A, Type{B})`](@ref) is true if and only if `A` and `B` are the same object
```

rfourquet commented on doc/src/manual/types.md:
```diff
@@ -1127,6 +1058,146 @@ must always be specified in full, this is not especially helpful, but in Julia,
+In other words, [`isa(A,Type{B})`](@ref) is true if and only if `A` and `B` are the same object
+and that object is a type.
+
+In particular, since parametric types are [invariant](@ref man-parametric-composite-types),
```
Sounds like the sentence is missing a "connection" word, e.g. I would expect "we have:" or something like that at the end, before the example.

rfourquet commented on doc/src/manual/types.md:
```diff
@@ -1127,6 +1058,146 @@ must always be specified in full, this is not especially helpful, but in Julia,
+While `Type` is part of Julia's type hierarchy like any other abstract parametric type, it
+is not commonly used outside method signatures except in some special cases. Another
+important use case for `Type` is sharpening field types which would otherwise be captured
+less precisely, eg as [`DataType`](@ref man-declared-types) in the example below where the
```
```suggestion
less precisely, e.g. as [`DataType`](@ref man-declared-types) in the example below where the
```

tpapp committed:
Wording corrections suggested by @rfourquet (thanks!)

tpapp committed:
remove "the"

Co-authored-by: Rafael Fourquet <fourquet.rafael@gmail.com>

tpapp committed:
spacing

Co-authored-by: Rafael Fourquet <fourquet.rafael@gmail.com>

tpapp committed:
e.g.

Co-authored-by: Rafael Fourquet <fourquet.rafael@gmail.com>

tpapp commented:
@rfourquet, thanks for the thorough reading and the detailed suggestions, I applied all of them.

StefanKarpinski commented:
Great clarification. Thank you! This looks ready to merge once CI passes.

dkarrasch commented on doc/src/manual/types.md:
```diff
@@ -1127,6 +1058,146 @@ must always be specified in full, this is not especially helpful, but in Julia,
 
+## [Singleton types](@id man-singleton-types)
+
+Immutable composite types with no fields are called *singletons*. Formally, if 
```
```suggestion
Immutable composite types with no fields are called *singletons*. Formally, if
```

dkarrasch committed:
fix whitespace

rfourquet commented on doc/src/manual/types.md:
```diff
@@ -1127,6 +1058,146 @@ must always be specified in full, this is not especially helpful, but in Julia,
+1. `T` is an immutable composite type (ie defined with `struct`),
+1. `a isa T && b isa T` implies `a === b`,
+
+then `T` is a singleton type.[^2] [`Base.issingletontype`](@ref) can be used to check if a
```
Is there an official convention on whether the footnote sign should go before or after the period? I saw both in the manual.

tpapp replied:
The [Chicago Manual of Style](https://www.chicagomanualofstyle.org/qanda/data/faq/topics/Punctuation/faq0020.html) suggests that it goes after the period.

tpapp replied:
Incidentally, since these questions are in general important for high-quality documentation, I started some discussion about referring a style guide in #36723.

rfourquet commented on doc/src/manual/types.md:
```diff
@@ -1127,6 +1058,146 @@ must always be specified in full, this is not especially helpful, but in Julia,
+
+Immutable composite types with no fields are called *singletons*. Formally, if
+
+1. `T` is an immutable composite type (ie defined with `struct`),
```
```suggestion
1. `T` is an immutable composite type (i.e. defined with `struct`),
```

rfourquet reviewed (approved):
Yes nice change! Just one more possible change ("ie" -> "i.e.") if you also update this branch for something else.

tpapp committed:
Update doc/src/manual/types.md

Co-authored-by: Rafael Fourquet <fourquet.rafael@gmail.com>

tpapp commented:
Another friendly bump: I think I addressed all comments.

I am happy to wait for more feedback if necessary, just don't want this PR to be forgotten.

tkf commented:
It has two approvals and the `merge me` label. I think it's more than enough :)
